File: ChessLab/Puzzles/PuzzleSolutionTrimmer.swift
import ChessKit

/// Tronque la variante principale (PV) d'un puzzle **issu d'une partie de
/// l'utilisateur** en une solution courte et nette.
///
/// La génération (`AnalysisViewModel.generatePuzzles`) part de la PV
/// complète du moteur (jusqu'à ~10 demi-coups). Exiger cette séquence
/// exacte est fragile : la queue d'une PV n'est pas fiable, et un coup
/// gagnant ALTERNATIF y serait compté faux (voir instructions.md §G7). Les
/// puzzles Lichess, eux, sont curés pour l'unicité — cette troncature ne
/// concerne QUE les puzzles maison.
///
/// Heuristique **purement matérielle** (aucun moteur requis, donc
/// déterministe et testable) : on rejoue la PV et on coupe dès qu'après un
/// coup du résolveur son avantage matériel devient décisif (`≥ +3`, un
/// coup gagnant de pièce mineure) ET stable (la riposte adverse forcée ne
/// le résorbe pas) — ou à un mat. Plafond dur à `maxPlies`. En pratique la
/// plupart des puzzles maison deviennent 1 à 3 coups. La solution se
/// termine toujours sur un coup du résolveur (longueur impaire), pour que
/// le dernier geste demandé à l'utilisateur soit SON coup.
enum PuzzleSolutionTrimmer {
    static func trim(
        pv: [String],
        startFEN: String,
        decisiveGain: Int = 3,
        maxPlies: Int = 6
    ) -> [String] {
        guard let start = Position(fen: startFEN) else {
            return endingOnSolverMove(Array(pv.prefix(maxPlies)))
        }
        let solverColor = start.sideToMove
        let baseDiff = materialDiff(start, from: solverColor)

        var board = Board(position: start)
        var applied: [String] = []

        for (ply, lan) in pv.prefix(maxPlies).enumerated() {
            guard apply(lan: lan, to: &board) else { break }
            applied.append(lan)

            if case .checkmate = board.state {
                return applied // mat : décisif quel que soit le matériel
            }

            let solverJustMoved = ply.isMultiple(of: 2) // pv[0] = coup du résolveur
            guard solverJustMoved else { continue }

            let gain = materialDiff(board.position, from: solverColor) - baseDiff
            guard gain >= decisiveGain else { continue }

            // Stabilité : la riposte forcée (ply+1) ne doit pas ramener
            // l'avantage sous le seuil (sinon c'est un simple échange en
            // cours, la tactique n'est pas encore résolue).
            let replyIndex = ply + 1
            if replyIndex < pv.count && replyIndex < maxPlies {
                var probe = board
                if apply(lan: pv[replyIndex], to: &probe) {
                    let gainAfterReply = materialDiff(probe.position, from: solverColor) - baseDiff
                    if gainAfterReply >= decisiveGain {
                        return applied // décisif et stable → coupe ici
                    }
                    continue // reprise adverse : l'échange continue
                }
            }
            return applied // pas de riposte (ou inapplicable) : déjà décisif
        }

        return endingOnSolverMove(applied)
    }

    /// Une solution doit se terminer sur un coup du résolveur (index pair,
    /// donc longueur impaire) : sinon le dernier demi-coup est une riposte
    /// adverse auto-jouée, ce qui n'a pas de sens comme état "résolu".
    private static func endingOnSolverMove(_ moves: [String]) -> [String] {
        guard moves.count > 1, moves.count.isMultiple(of: 2) else { return moves }
        return Array(moves.dropLast())
    }

    /// Somme des valeurs matérielles de `color` moins celle de l'adversaire
    /// (le roi vaut 0, il n'entre pas dans le calcul).
    private static func materialDiff(_ position: Position, from color: Piece.Color) -> Int {
        position.pieces.reduce(0) { total, piece in
            let value = pieceValue(piece.kind)
            return total + (piece.color == color ? value : -value)
        }
    }

    /// Applique un coup en notation LAN ("e2e4", "e7e8q") sur `board`,
    /// promotion comprise. Même logique que
    /// `PuzzleSolveViewModel.applyForcedMove`.
    private static func apply(lan: String, to board: inout Board) -> Bool {
        guard lan.count >= 4 else { return false }
        let start = Square(String(lan.prefix(2)))
        let end = Square(String(lan.dropFirst(2).prefix(2)))
        guard let move = board.move(pieceAt: start, to: end) else { return false }
        if case .promotion = board.state {
            let kind: Piece.Kind = lan.count == 5
                ? (Piece.Kind(rawValue: String(lan.suffix(1)).uppercased()) ?? .queen)
                : .queen
            _ = board.completePromotion(of: move, to: kind)
        }
        return true
    }
}

File: android/app/src/main/java/com/chesslab/puzzles/PuzzleSolutionTrimmer.kt
package com.chesslab.puzzles

import chesskit.Board
import chesskit.Piece
import chesskit.Position
import chesskit.Square
import com.chesslab.play.CapturedMaterial

/**
 * Tronque la variante du moteur en une solution COURTE ET NETTE, pour les
 * puzzles tirés d'une partie de l'utilisateur. Pendant de
 * `PuzzleSolutionTrimmer.swift`.
 *
 * Exiger la séquence complète du moteur est fragile : la queue d'une variante
 * n'est pas fiable, et un coup gagnant ALTERNATIF y serait compté faux. Les
 * puzzles Lichess, eux, sont curés pour l'unicité — cette troncature ne
 * concerne QUE les puzzles maison.
 *
 * L'heuristique est purement MATÉRIELLE, donc sans moteur : déterministe, et
 * vérifiable sur des positions écrites à la main. On rejoue la variante et on
 * coupe dès qu'après un coup du résolveur son avantage devient décisif (une
 * pièce mineure) ET stable — ou à un mat.
 */
object PuzzleSolutionTrimmer {

    fun trim(
        pv: List<String>,
        startFen: String,
        decisiveGain: Int = 3,
        maxPlies: Int = 6,
    ): List<String> {
        val start = Position.fromFen(startFen)
            ?: return endingOnSolverMove(pv.take(maxPlies))
        val solver = start.sideToMove
        val baseDiff = materialDiff(start, solver)

        val board = Board(start)
        val applied = ArrayList<String>()

        pv.take(maxPlies).forEachIndexed { ply, lan ->
            if (!apply(lan, board)) return endingOnSolverMove(applied)
            applied += lan

            // Un mat est décisif quel que soit le matériel.
            if (board.state is Board.State.Checkmate) return applied

            // pv[0] est le coup du résolveur : il joue les index pairs.
            if (ply % 2 != 0) return@forEachIndexed

            val gain = materialDiff(board.position, solver) - baseDiff
            if (gain < decisiveGain) return@forEachIndexed

            // Stabilité : la riposte forcée ne doit pas résorber l'avantage,
            // sinon c'est un échange en cours et la tactique n'est pas résolue.
            val replyIndex = ply + 1
            if (replyIndex < pv.size && replyIndex < maxPlies) {
                val probe = Board(board.position.copy())
                if (apply(pv[replyIndex], probe)) {
                    val after = materialDiff(probe.position, solver) - baseDiff
                    if (after >= decisiveGain) return applied
                    return@forEachIndexed
                }
            }
            return applied
        }
        return endingOnSolverMove(applied)
    }

    /**
     * Une solution se termine toujours sur un coup du RÉSOLVEUR (longueur
     * impaire) : sinon le dernier demi-coup est une riposte auto-jouée, ce qui
     * n'a pas de sens comme état « résolu ».
     */
    private fun endingOnSolverMove(moves: List<String>): List<String> =
        if (moves.size > 1 && moves.size % 2 == 0) moves.dropLast(1) else moves

    /** Le matériel de [color] moins celui de l'adversaire (le roi vaut 0). */
    private fun materialDiff(position: Position, color: Piece.Color): Int =
        position.pieces.sumOf {
            val v = CapturedMaterial.value(it.kind)
            if (it.color == color) v else -v
        }

    private fun apply(lan: String, board: Board): Boolean {
        if (lan.length < 4) return false
        val move = board.move(Square(lan.substring(0, 2)), Square(lan.substring(2, 4))) ?: return false
        if (board.state is Board.State.Promotion) {
            val kind = when (lan.getOrNull(4)?.lowercaseChar()) {
                'n' -> Piece.Kind.knight
                'b' -> Piece.Kind.bishop
                'r' -> Piece.Kind.rook
                else -> Piece.Kind.queen
            }
            board.completePromotion(move, kind)
        }
        return true
    }
}
